crates/sequencer/src/sequencer.rs

```diff
@@ -591,94 +597,153 @@ where
+                    .await
+                    .map_err(|_| anyhow!("DA service is dead!"))?
+                    .map_err(|_| anyhow!("Send transaction cannot fail"))?;
+
+                ledger_db
+                    .set_last_sequencer_commitment_l2_height(l2_end)
+                    .map_err(|_| {
+                        anyhow!("Sequencer: Failed to set last sequencer commitment L2 height")
+                    })?;
+
+                debug!("Commitment info: {:?}", commitment_info);
+
+                if let Some(db_config) = db_config {
+                    match PostgresConnector::new(db_config).await {
+                        Ok(pg_connector) => {
+                            pg_connector
+                                .insert_sequencer_commitment(
+                                    Into::<[u8; 32]>::into(tx_id).to_vec(),
+                                    l2_start.0 as u32,
+                                    l2_end.0 as u32,
+                                    commitment.merkle_root.to_vec(),
+                                    CommitmentStatus::Mempool,
+                                )
+                                .await
+                                .map_err(|_| {
+                                    anyhow!("Sequencer: Failed to insert sequencer commitment")
+                                })?;
+                        }
+                        Err(e) => {
+                            warn!("Failed to connect to postgres: {:?}", e);
+                        }
                     }
                 }
+
+                // Remove commitment from pending commitments
+                let mut pending_commitments = pending_commitments_l2_range.lock().await;
+                pending_commitments.retain(|&(start, end)| start != l2_start || end != l2_end);
+                ledger_db.set_pending_commitments_l2_range(&pending_commitments)?;
+
+                info!("New commitment. L2 range: #{}-{}", l2_start.0, l2_end.0);
+                Ok(())
             }
+            .await;
 
-            // Clear state diff.
+            if let Err(err) = result {
+                error!(
+                    "Error in spawned task for handling commitment result: {}",
+                    err
+                );
+            }
+        };
+
+        if !resubmit {
+            // Add commitment to pending commitments
+            let mut pending_commitments = self.pending_commitments_l2_range.lock().await;
+            pending_commitments.push((*l2_range_to_submit.start(), *l2_range_to_submit.end()));
+            self.ledger_db
+                .set_pending_commitments_l2_range(&pending_commitments)?;
+
+            // Clear state diff
             self.ledger_db.set_state_diff(vec![])?;
             self.last_state_diff = vec![];
 
-            info!("New commitment. L2 range: #{}-{}", l2_start, l2_end,);
+            // Handle DA response asynchronously
+            tokio::spawn(handle_da_response);
```

Comment:
Here we're removing the current commitment from pending. However, `handle_da_response` here would be called again and we would remove the same commitment from the list again which has already been removed. I also don't see how calling it again would process other commitments?

Reply:
AFAIK, `handle_da_response`, which is a Future, will not execute until await is called on it. Hence we are actually not calling it twice, but we are either handing the future to tokio to handle it in a separate green thread (case of regular submission), or we are blocking further execution by polling the Future within the current thread.

I believe this should also answer the previous conversation about oneshot channels.